kore/src/Kore/Builtin/AssociativeCommutative.hs

```diff
@@ -0,0 +1,1451 @@
+
+The domain values are assumed to have the same sort, but this is not checked. If
+multiple sorts are hooked to the same builtin domain, the verifier should
+reject the definition.
+-}
+unifyEqualsNormalized
+    :: forall normalized unifier (valueWrapper :: * -> *) variable
+    .   ( SortedVariable variable
+        , Unparse variable
+        , Show variable
+        , Traversable valueWrapper
+        , FreshVariable variable
+        , TermWrapper normalized valueWrapper
+        , MonadUnify unifier
+        , Unparse (valueWrapper (TermLike variable))
+        )
+    => SmtMetadataTools Attribute.Symbol
+    -> TermLike variable
+    -> TermLike variable
+    -> (TermLike variable -> TermLike variable -> unifier (Pattern variable))
+    -> Bool
+    -> Domain.InternalAc (TermLike Concrete) normalized (TermLike variable)
+    -> Domain.InternalAc (TermLike Concrete) normalized (TermLike variable)
+    -> MaybeT unifier (Pattern variable)
+unifyEqualsNormalized
+    tools
+    first
+    second
+    unifyEqualsChildren
+    alreadyNormalized
+    normalized1
+    normalized2
+  = do
+    let
+        Domain.InternalAc { builtinAcChild = firstWrapped } =
+            normalized1
+        Domain.InternalAc { builtinAcChild = secondWrapped } =
+            normalized2
+        firstNormalized = Domain.unwrapAc firstWrapped
+        secondNormalized = Domain.unwrapAc secondWrapped
+
+    unifierNormalized <-
+        unifyEqualsNormalizedAc
+            tools
+            first
+            second
+            unifyEqualsChildren
+            alreadyNormalized
+            firstNormalized
+            secondNormalized
+    let
+        unifierNormalizedTerm :: TermNormalizedAc valueWrapper variable
+        unifierPredicate :: Predicate variable
+        (unifierNormalizedTerm, unifierPredicate) =
+            Conditional.splitTerm unifierNormalized
+        normalizedTerm :: TermLike variable
+        normalizedTerm = asInternal tools sort1 unifierNormalizedTerm
+
+    -- TODO(virgil): Check whether this normalization is still needed
+    -- (the normalizedTerm may already be re-normalized) and remove it if not.
+    renormalized <- normalize1 normalizedTerm
+
+    let unifierTerm :: TermLike variable
+        unifierTerm = asInternal tools sort1 renormalized
+    return (unifierTerm `withCondition` unifierPredicate)
+  where
+    sort1 = termLikeSort first
+
+    normalize1
+        ::  ( MonadUnify unifier
+            , Ord variable
+            )
+        => TermLike variable
+        -> MaybeT unifier (TermNormalizedAc valueWrapper variable)
+    normalize1 patt =
+        case toNormalized tools patt of
+            Bottom -> Monad.Trans.lift $ Monad.Unify.explainAndReturnBottom
+                "Duplicated elements in normalization."
+                first
+                second
+            Normalized n -> return n
+
+{- | Unifies two AC structs represented as @NormalizedAc@.
```

Comment:
this seems really, really long.  Would it make sense to split it intro smaller functions?

Reply:
This is not trivial to split, and other people seem to be waiting for this PR, so I added a TODO and I'll try to fix it later. I'll remove the "Fixes" pr description.